File: apps/api/models/project.py
import enum
from datetime import datetime, timezone
from decimal import Decimal
from sqlalchemy import String, Integer, Numeric, DateTime, Date, ForeignKey, func, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from db.session import Base


class ProjectStatus(str, enum.Enum):
    PENDING_ADVANCE = "PENDING_ADVANCE"
    IN_PROGRESS = "IN_PROGRESS"
    IN_REVIEW = "IN_REVIEW"
    FINISHED = "FINISHED"
    PENDING_FINAL = "PENDING_FINAL"
    PAID = "PAID"
    DELIVERED = "DELIVERED"


def _now() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


class Project(Base):
    __tablename__ = "projects"

    id: Mapped[str] = mapped_column(String, primary_key=True)
    code: Mapped[str] = mapped_column(String, unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[str | None] = mapped_column(String, nullable=True)
    status: Mapped[str] = mapped_column(String, default="PENDING_ADVANCE")
    total_value: Mapped[Decimal] = mapped_column("total_value", Numeric(18, 2), nullable=False)
    advance_percent: Mapped[int] = mapped_column("advance_percent", Integer, nullable=False)
    progress: Mapped[int] = mapped_column(Integer, default=0)
    start_date: Mapped[datetime | None] = mapped_column("start_date", DateTime(timezone=True), nullable=True)
    estimated_date: Mapped[datetime | None] = mapped_column("estimated_date", DateTime(timezone=True), nullable=True)
    delivered_at: Mapped[datetime | None] = mapped_column("delivered_at", DateTime(timezone=True), nullable=True)
    drive_folder_id: Mapped[str | None] = mapped_column("drive_folder_id", String, nullable=True)
    client_id: Mapped[str] = mapped_column("client_id", String, ForeignKey("users.id"), nullable=False)
    created_at: Mapped[datetime] = mapped_column("created_at", DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column("updated_at", DateTime, default=_now, onupdate=_now)

    members: Mapped[list["ProjectMember"]] = relationship("ProjectMember", back_populates="project", lazy="selectin")


class ProjectMember(Base):
    __tablename__ = "project_members"
    __table_args__ = (UniqueConstraint("project_id", "user_id"),)

    project_id: Mapped[str] = mapped_column("project_id", String, ForeignKey("projects.id"), nullable=False, primary_key=True)
    user_id: Mapped[str] = mapped_column("user_id", String, ForeignKey("users.id"), nullable=False, primary_key=True)

    project: Mapped["Project"] = relationship("Project", back_populates="members")

File: apps/web/src/types/project.ts
export type ProjectStatus =
  | "PENDING_ADVANCE"
  | "IN_PROGRESS"
  | "IN_REVIEW"
  | "FINISHED"
  | "PENDING_FINAL"
  | "PAID"
  | "DELIVERED";

export interface Project {
  id: string;
  code: string;
  name: string;
  description: string | null;
  status: ProjectStatus;
  total_value: string;
  advance_percent: number;
  progress: number;
  start_date: string | null;
  estimated_date: string | null;
  delivered_at: string | null;
  drive_folder_id: string | null;
  client_id: string;
  client_name: string;
  client_email: string;
  member_ids: string[];
  member_names: string[];
  created_at: string;
  updated_at: string;
}

export const STATUS_LABELS: Record<ProjectStatus, string> = {
  PENDING_ADVANCE: "Pendiente anticipo",
  IN_PROGRESS: "En ejecución",
  IN_REVIEW: "En revisión",
  FINISHED: "Finalizado",
  PENDING_FINAL: "Pendiente pago final",
  PAID: "Pagado",
  DELIVERED: "Entregado",
};

export const STATUS_COLORS: Record<ProjectStatus, string> = {
  PENDING_ADVANCE: "bg-yellow-100 text-yellow-800",
  IN_PROGRESS: "bg-blue-100 text-blue-800",
  IN_REVIEW: "bg-purple-100 text-purple-800",
  FINISHED: "bg-green-100 text-green-800",
  PENDING_FINAL: "bg-orange-100 text-orange-800",
  PAID: "bg-emerald-100 text-emerald-800",
  DELIVERED: "bg-gray-100 text-gray-800",
};
